src/Xunit.StaFact/Sdk/UIFactDiscoverer.cs

```diff
@@ -20,8 +22,103 @@ public UIFactDiscoverer(IMessageSink diagnosticMessageSink)
         this.diagnosticMessageSink = diagnosticMessageSink;
     }
 
+    internal static UISettingsKey GetSettings(ITestMethod testMethod, IAttributeInfo factAttribute)
+    {
+        var maxAttempts = GetMaxAttempts(testMethod, factAttribute);
+        return new UISettingsKey(maxAttempts);
+    }
+
     protected override IXunitTestCase CreateTestCase(ITestFrameworkDiscoveryOptions discoveryOptions, ITestMethod testMethod, IAttributeInfo factAttribute)
     {
-        return new UITestCase(UITestCase.SyncContextType.Portable, this.diagnosticMessageSink, discoveryOptions.MethodDisplayOrDefault(), testMethod);
+        var maxAttempts = GetMaxAttempts(testMethod, factAttribute);
+        var settings = new UISettingsKey(maxAttempts);
+        return new UITestCase(UITestCase.SyncContextType.Portable, this.diagnosticMessageSink, discoveryOptions.MethodDisplayOrDefault(), testMethod, testMethodArguments: null, settings);
+    }
+
+    private static int GetMaxAttempts(ITestMethod testMethod, IAttributeInfo factAttribute)
+    {
+        return GetMaxAttempts(factAttribute, GetSettingsAttributes(testMethod).ToArray());
+    }
+
+    private static IEnumerable<IAttributeInfo> GetSettingsAttributes(ITestMethod testMethod)
+    {
+        foreach (IAttributeInfo attributeData in testMethod.Method.GetCustomAttributes(typeof(UISettingsAttribute)))
+        {
+            yield return attributeData;
+        }
+
+        foreach (IAttributeInfo attributeData in testMethod.TestClass.Class.GetCustomAttributes(typeof(UISettingsAttribute)))
+        {
+            yield return attributeData;
+        }
+    }
+
+    private static int GetMaxAttempts(IAttributeInfo factAttribute, IAttributeInfo[] settingsAttributes)
+    {
+        return GetNamedArgument(
+            factAttribute,
+            settingsAttributes,
+            nameof(UISettingsAttribute.MaxAttempts),
+            static value => value > 0,
+            defaultValue: 1);
+    }
+
+    private static TValue GetNamedArgument<TValue>(IAttributeInfo factAttribute, IAttributeInfo[] settingsAttributes, string argumentName, Func<TValue, bool> isValidValue, TValue defaultValue)
+    {
+        return GetNamedArgument(
+            factAttribute,
+            settingsAttributes,
+            argumentName,
+            isValidValue,
+            merge: null,
+            defaultValue);
+    }
+
+    private static TValue GetNamedArgument<TValue>(IAttributeInfo factAttribute, IAttributeInfo[] settingsAttributes, string argumentName, Func<TValue, bool> isValidValue, Func<TValue, TValue, TValue>? merge, TValue defaultValue)
+    {
+        StrongBox<TValue>? result = null;
+        if (TryGetNamedArgument(factAttribute, argumentName, isValidValue, out TValue? value))
```

Comment:
📝 This was originally designed around having the properties be settable on both the fact attribute (e.g. `WinFormsFactAttribute.MaxAttempts`) and a separate settings attribute (in this case, `UISettingsAttribute`). I retained the full logic even though the current implementation only allows the values to be set in one location.